app/grpc/gasestimation/gas_estimator.go

```diff
@@ -0,0 +1,246 @@
+	// estimate the gas price
+	gasPrice, err := s.estimateGasPrice(ctx, request.TxPriority)
+	if err != nil {
+		return nil, err
+	}
+
+	// estimate the gas used
+	btx, isBlob, err := blobtx.UnmarshalBlobTx(request.TxBytes)
+	if isBlob && err != nil {
+		return nil, err
+	}
+
+	var txBytes []byte
+	if isBlob {
+		txBytes = btx.Tx
+	} else {
+		txBytes = request.TxBytes
+	}
+
+	gasUsedInfo, _, err := s.simulateFn(txBytes)
+	if err != nil {
+		return nil, err
+	}
+	return &EstimateGasPriceAndUsageResponse{
+		EstimatedGasPrice: gasPrice,
+		EstimatedGasUsed:  gasUsedInfo.GasUsed,
+	}, nil
+}
+
+// estimateGasPrice takes a transaction priority and estimates the gas price based
+// on the gas prices of the transactions in the last five blocks.
+// If no transaction is found in the last five blocks, return the network
+// min gas price.
+// It's up to the light client to set the gas price in this case
+// to the minimum gas price set by that node.
+func (s *gasEstimatorServer) estimateGasPrice(ctx context.Context, priority TxPriority) (float64, error) {
+	status, err := s.clientCtx.Client.Status(ctx)
+	if err != nil {
+		return 0, err
+	}
+	latestHeight := status.SyncInfo.LatestBlockHeight
+	page := 1
+	txSearchResult, err := s.clientCtx.Client.TxSearch(
+		ctx,
+		lastFiveBlocksTransactionsQuery(latestHeight),
+		false,
+		&page,
+		&numberOfTransactionsPerPage,
+		"asc",
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	totalNumberOfTransactions := txSearchResult.TotalCount
+	if totalNumberOfTransactions == 0 {
+		// return the min gas price if no transaction found in the last 5 blocks
+		return minfee.DefaultNetworkMinGasPrice.MustFloat64(), nil
+	}
+
+	gasPrices := make([]float64, 0)
+	for {
+		currentPageGasPrices, err := extractGasPriceFromTransactions(txSearchResult.Txs)
+		if err != nil {
+			return 0, err
+		}
+		gasPrices = append(gasPrices, currentPageGasPrices...)
+		if len(gasPrices) >= totalNumberOfTransactions {
+			break
+		}
+		page++
+		txSearchResult, err = s.clientCtx.Client.TxSearch(
+			ctx,
+			lastFiveBlocksTransactionsQuery(latestHeight),
+			false,
+			&page,
+			&numberOfTransactionsPerPage,
+			"asc",
+		)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return estimateGasPriceForTransactions(gasPrices, priority)
+}
+
+// estimateGasPriceForTransactions takes a list of transactions and priority
+// and returns a gas price estimation.
+// The priority sets the estimation as follows:
+// - High Priority: The gas price is the price at the start of the top 10% of transactions’ gas prices from the last five blocks.
+// - Medium Priority: The gas price is the median of all gas prices from the last five blocks.
+// - Low Priority: The gas price is the value at the end of the lowest 10% of gas prices from the last five blocks.
+// - Unspecified Priority (default): This is equivalent to the Medium priority, using the median of all gas prices from the last five blocks.
+// More information can be found in ADR-023.
```

Comment:
nit this comment says median but the implementation say mean.

```suggestion
// estimateGasPriceForTransactions takes a list of transactions and priority
// and returns a gas price estimation.
// The priority sets the estimation as follows:
// - High Priority: The gas price is the price at the start of the top 10% of transactions’ gas prices from the last five blocks.
// - Medium Priority: The gas price is the mean of all gas prices from the last five blocks.
// - Low Priority: The gas price is the value at the end of the lowest 10% of gas prices from the last five blocks.
// - Unspecified Priority (default): This is equivalent to the Medium priority, using the mean of all gas prices from the last five blocks.
// More information can be found in ADR-023.
```